planet/api/models.py

```diff
@@ -180,105 +134,277 @@ def name(self):
+        async for c in self.response.aiter_bytes():
+            yield c
+
+    async def write(self, filename, overwrite=True, progress_bar=True):
+        '''Write the body to a file.
+
+        :param filename: Name to assign to downloaded file.
+        :type filename: str
+        :param overwrite: Overwrite any existing files. Defaults to True
+        :type overwrite: boolean, optional
+        :param progress_bar: Show progress bar during download. Defaults to
+            True.
+        :type progress_bar: boolean, optional
         '''
-        if not file:
-            file = self.name
-        if not file:
-            raise ValueError('no file name provided or discovered in response')
-        if hasattr(file, 'write'):
-            self._write(file, callback)
-        else:
-            with atomic_open(file, 'wb') as fp:
-                self._write(fp, callback)
-
-    def write_to_file(self, filename=None, overwrite=True, callback=None):
-        '''Write the contents of the body to the optionally provided filename.
-
-        providing progress to the optional callback. The callback will be
-        invoked 3 different ways:
-
-        * First as ``callback(start=self)``
-        * For each chunk of data written as
-          ``callback(wrote=chunk_size_in_bytes, total=all_byte_cnt)``
-        * Upon completion as ``callback(finish=self)`
-        * Upon skip as `callback(skip=self)`
-
-        :param filename: Filename to write to, defaults to body name
-        :type filename: str, optional
-        :param overwrite: Specify whether the file at filename
-            should be overwritten if it exists, defaults to True
-        :type overwrite: bool, optional
-        :param callback: A function handle of the form
-            ``callback(start, wrote, total, finish, skip)`` that receives write
-            progress. Defaults to None
-        :type callback: function, optional
+        class _LOG():
+            def __init__(self, total, unit, filename, disable):
+                self.total = total
+                self.unit = unit
+                self.disable = disable
+                self.previous = 0
+                self.filename = filename
+
+                if not self.disable:
+                    LOGGER.debug(f'writing to {self.filename}')
+
+            def update(self, new):
+                if new-self.previous > self.unit and not self.disable:
+                    # LOGGER.debug(f'{new-self.previous}')
+                    perc = int(100 * new / self.total)
+                    LOGGER.debug(f'{self.filename}: '
+                                 f'wrote {perc}% of {self.total}')
+                    self.previous = new
+
+        unit = 1024*1024
+
+        mode = 'wb' if overwrite else 'xb'
+        try:
+            with open(filename, mode) as fp:
+                _log = _LOG(self.size, 16*unit, filename, disable=progress_bar)
+                with tqdm(total=self.size, unit_scale=True,
```

Comment:
This is a red flag to me along with `_LOG` and I'm curious why TTY-emitting functionality is in the response model's body's write method?

Reply:
I agree, I wasn't happy with the reporting logic being within the write method either. We want to achieve real-time reporting without having to use threading. Real-time reporting means it has to be done within the context of the open file. In sync code (e.g. v1) this was achieved by defining a reporting function elsewhere, and passing into this function as a callback. I'm not certain how callbacks and async behave together. So this has the benefit of being simple. It has the drawback of being somewhat 'polluted'. Given that we can implement an async callback, is that any better than providing reporting within the write function? What implementation would you propose?

Here's the inspiration for the current tqdm usage (from httpx, which we depend upon here): https://www.python-httpx.org/advanced/#monitoring-download-progress

@sarasafavi it would be great to get your take on this as well

Reply:
As far as I'm aware, putting view code in a model is an anti-pattern and is widely and strongly discouraged. Additionally, I'm concerned about the violation of the Principle of Least Astonishment I see immediately as a developer (i.e. I think it would be widely-accepted as surprising to have a `session.request(...)` or `session.stream(...)` try and render progress bars). I haven't run tests/timeit but I'm also concerned about the performance of this at scale as well as the potential for unintentional dark patterns creeping up like Hidden Costs. For an implementation I would suggest to have all the console-ish responsibilities in the CLI.

Reply:
I agree that it is a bit muddy (polluted, as @jreiberkyle said) but it's certainly not a red flag or anti-pattern. If future development discovers significant timing issues we can revisit it. At this point in development though this is fine as-is.